### PR TITLE
Remove dependency on the system graphviz when rendering crate graph

### PR DESCRIPTION
**Pull request by p32blo**
This PR removes the need for having `graphviz` installed on the user system by using the `d3-graphviz` npm package.

The responsibility of rendering the svg output is moved to the extension while the rust side only handles the generation of the dot file.

This change also brings the following additional features:
- Allow zooming the view
- Ctrl+click to reset the zoom 
- Adjust the color scheme to dark themes
- Works on any platform without installing graphviz locally

---

I’m not sure if this fits what you had in mind for the crates graph feature but I decided to submit it anyway to see if this is useful to anyone else. 

A potential downside might be that it increases the extension size ( haven’t checked) but this feature already required the installation of graphviz on the user side, so the cost is just moved explicitly to the extension.

Feel free to make any suggestion or comments.

### CODE REVIEW

**Comment by lnicola** [comment]
There's a formatting difference in the Rust code and some TS lints too.

File: editors/code/src/commands.ts [hunk 1]
```diff
@@ -481,12 +482,67 @@ export function viewItemTree(ctx: Ctx): Cmd {
...
+        });
         const params = {
             full: full,
         };
-        const svg = await ctx.client.sendRequest(ra.viewCrateGraph, params);
-        panel.webview.html = svg;
+
+        const dot = await ctx.client.sendRequest(ra.viewCrateGraph, params);
+
+        const scripts = [
```
Comment (lnicola) [inline editors/code/src/commands.ts:498]: This seems a bit overkill. Your choice, but I'd probably use `${nodeModulesPath}` directly in the HTML template below.

**Comment by lnicola** [comment]
r? @matklad 

I know this brings some extra dependencies on the JS side, but:

 - they're not build (bundler) -related, so they probably behave better on updates
 - a lot of people complained about having to install GraphViz (or at least about the confusing error message when they didn't have it)
 - it's still something for non-Code users: if you get the `dot` file you can paste it yourself somewhere and run GraphViz, and it's a bit nicer to edit
 - the accessibility improvements are nice and were requested by users, though we could implement those over a plain SVG too.

**Comment by lnicola** [comment]
@p32blo did you actually test this in Code (using `cargo xtask install`)? It doesn't work for me, because you're loading paths that don't exist in the installed extension. There's no `node_modules`, only `out`, and it doesn't have `d3`. You might need to tweak `.vscodeignore` and maybe look into how you can bundle extra files inside an extension.

**Comment by lnicola** [comment]
This also needs a rebase, another PR touched `package-lock.json`.

**Comment by matklad** [comment]
Yeah, seems like a better factoring! 

Question: how much larger is the vsix? Looks like this uses graphviz compiled to wasm, I wonder if that makes the overall download significantly larger. 

When we merge this, we probably need to ping the #4604

**Comment by bjorn3** [comment]
The total node_modules is 9.3MB when installing just d3-graphviz and 18MB when installing d3 too. Just the necessary files is ~1.6MB:

```
$ du -sh node_modules/d3/dist/d3.min.js
268K    node_modules/d3/dist/d3.min.js
$ du -sh node_modules/@hpcc-js/wasm/dist/{index.min.js,*.wasm}
112K    node_modules/@hpcc-js/wasm/dist/index.min.js
100K    node_modules/@hpcc-js/wasm/dist/expatlib.wasm
1,1M    node_modules/@hpcc-js/wasm/dist/graphvizlib.wasm
$ du -sh node_modules/d3-graphviz/build/d3-graphviz.min.js
32K     node_modules/d3-graphviz/build/d3-graphviz.min.js
```

**Comment by lnicola** [comment]
The `vsix` is probably smaller than that because it's compressed.

Right now it's not larger because it's actually missing the new files, of course.

**Comment by p32blo** [comment]
Thank you everyone for the comments! I updated the PR to fix some of the issues.

> @p32blo did you actually test this in Code (using `cargo xtask install`)?

I only ran the 'Run Extension (Debug Build)' task to test the changes. I added each required file explicitly to `.vsignore` and testing with `cargo xtask install` and it seems to work now. Can you check if it works for you now?


> This also needs a rebase, another PR touched `package-lock.json`.

I rebased to master in this last update.  


> (not sure if expatlib.wasm is necessary or not)

Only the `graphvizlib.wasm` is necessary.

File: editors/code/src/commands.ts (hunk 1 above)
Comment (p32blo) [reply editors/code/src/commands.ts]: You are right. I tried to simplify this in the new PR update. Thank you.

**Comment by lnicola** [comment]
It works for me now, but how do I change to dark mode?

**Comment by p32blo** [comment]
It adapts to dark themes, here are some screenshots:

image
image

**Comment by lnicola** [comment]
Oh, right. I tried that, even noticed the white lines, but thought it wasn't working because of the white box backgrounds.

**Comment by lnicola** [comment]
So the VSIX is 1.1 MB, up from 310 KB. That's a hefty increase, but not so bad in comparison to the LSP binary (around 9 MB). So I'm inclined to:

bors r+

changelog feature (first contribution) use `graphviz-d3` for the crate graph instead of running an installed binary.

**Comment by p32blo** [comment]
> 
> 
> Oh, right. I tried that, even noticed the white lines, but thought it wasn't working because of the white box backgrounds. It's even written in the CSS.

Yeah, it's just the lines xD. This is just a proof of concept and I'm sure someone with more CSS experience can do a better job later :D

**Pull request merged**